### ISSUE TITLE
Confidence Interval

### ISSUE THREAD

**Issue by DoDzilla-ai** [dreamquark-ai/tabnet#320]
## Feature request
I would love to get a confidence interval feature for `TabNetRegressor`. Or even a tutorial notebook to calculate this would be great.

**What is the expected behavior?**
For a given alpha value, calculate the confidence interval (upper and lower limits).

**What is motivation or use case for adding/changing the behavior?**
In some cases, it is really hard to predict the exact numerical value. Instead, a range could be useful. 

**How should this be implemented in your opinion?**
I don't know tbh.

**Are you willing to work on this yourself?**
no

**Comment by eduardocarvp** [comment]
I'm not sure how we could implement this. Is there any chance another library like [mlinsights - IntervalRegressor](http://www.xavierdupre.fr/app/mlinsights/helpsphinx/mlinsights/mlmodel/interval_regressor.html#mlinsights.mlmodel.interval_regressor.IntervalRegressor) could be useful for you in this case?

**Comment by DoDzilla-ai** [comment]
@eduardocarvp So, I've tried your suggestion but I didn't understand much from the result. It doesn't seem like returning a range, but a single prediction (or predictions for each estimators).

**Comment by Optimox** [comment]
Hello @DoDzilla-ai,

This is an interesting topic, even-though there's nothing specific to TabNet in trying to compute confidence intervals. 

There is one method specific to deep learning that could be (quite) easily implemented with pytorch-tabnet in order to get confidence intervals.

Let's say you want 95% and 5% intervals for the example's sake.

To get the 95% interval predictions, you can train a model with a asymmetrical loss: for positive errors you penalize with L=x but for negative errors you penalize 1/0.05=20 times more with L=-20*x. This will shift your model predictions so that on average the model will predict something smaller than the actual value only 5% of the time, which is your 95% confidence interval predictions.

To get the 5% interval predictions, you do the opposite and penalize more the positive errors.

In order to use this with pytorch-tabnet here are the steps you need to follow:
- triplicate your targets so that your model will now have 3 outputs and not 1
- define a custom loss combining the loss you want to minimize for the main predictions (MAE, MSE etc...) and the 5% and 95% losses described above. (You'll have to give a weight to each loss, choosing the correct weights is up to you)
- train one single TabNet model that will learn to predict the actual predictions and the two confidence intervals

Then you'll have one model able to make predictions and give confidence intervals!

I think this method is quite easy to implement and is worth a shot. If you have troubles implementing it do not hesitate to ask for help here. I would be curious to know about your results.
